### ISSUE TITLE
MA0047 - Declare types in namespaces is handled by CA1050

### ISSUE THREAD

**Issue by Meir017** [meziantou/Meziantou.Analyzer#834]
- **Rule**: [MA0047](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0047.md) - Declare types in namespaces
- **Built-in Equivalent**: [CA1050](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1050) - Declare types in namespaces

## Analysis

**MA0047** and **CA1050** have **identical names and purposes**: ensuring that types are declared within namespaces rather than at the global scope, which is a fundamental .NET design guideline.

### MA0047 Coverage
- Detects types declared outside of namespaces
- Enforces proper namespace organization
- Promotes clean code architecture

### CA1050 Coverage  
- Detects types that should be declared in namespaces
- Official Microsoft rule for type design guidelines
- **Exact same name and functionality** as MA0047
- Part of the .NET Framework Design Guidelines

## Code Examples

**Problematic code detected by both rules:**
```csharp
// Both MA0047 and CA1050 will flag this
public class MyClass // No namespace declaration
{
    // ...
}

public enum MyEnum // No namespace declaration
{
    Value1,
    Value2
}
```

**Recommended fix:**
```csharp
// Both rules suggest organizing types in namespaces
namespace MyProject.Core
{
    public class MyClass
    {
        // ...
    }

    public enum MyEnum
    {
        Value1,
        Value2
    }
}
```

## Design Benefits

Declaring types in namespaces provides:
- **Name collision avoidance** across different assemblies
- **Logical organization** of related types
- **IntelliSense improvement** in IDEs
- **Assembly structure clarity**

## Recommendation

**CA1050** is the official Microsoft analyzer rule that provides **identical functionality** to MA0047, including the exact same name. Since both rules are functionally equivalent and CA1050 is the standard rule recognized across the .NET ecosystem, **MA0047 should be deprecated** in favor of CA1050.

## Migration Path

1. Enable **CA1050** in your analyzer configuration:
   ```xml
   <WarningsAsErrors Include="CA1050" />
   ```
2. Disable **MA0047**:
   ```xml
   <WarningsNotAsErrors Include="MA0047" />
   ```
3. Update any rule suppressions from MA0047 to CA1050

## Related

- Parent issue: #835 - Consider deprecating rules that are now covered by built-in .NET analyzers

**Comment by Meir017** [comment]
## Analysis: MA0047 vs CA1050 - Equivalent Rules ✅

After detailed analysis, **MA0047** and **CA1050** are **equivalent rules** with **identical names and purposes**.

### Equivalency Assessment

Both rules:
- Have the **exact same name**: "Declare types in namespaces"
- Detect types declared outside of namespaces (in global scope)
- Enforce fundamental .NET design guidelines for type organization
- Prevent name collisions and improve code organization

### Key Similarities

1. **Identical Name & Purpose**: Both rules are called "Declare types in namespaces"
2. **Same Target Pattern**: Both detect types in global scope that should be in namespaces
3. **Same Design Principle**: Both enforce the .NET Framework Design Guidelines
4. **Same Benefits**: Both prevent name collisions and improve code organization

### Example Code

```csharp
// Both MA0047 and CA1050 flag this:
public class MyClass // No namespace declaration
{
    // ...
}

// Both suggest this fix:
namespace MyProject.Core
{
    public class MyClass
    {
        // ...
    }
}
```

### Design Benefits

Declaring types in namespaces provides:
- **Name collision avoidance** across different assemblies
- **Logical organization** of related types
- **IntelliSense improvement** in IDEs
- **Assembly structure clarity**

### Recommendation: Deprecate MA0047 ✅

**CA1050** is the official Microsoft analyzer rule that provides **identical functionality** to MA0047, including the exact same name. Since both rules are functionally equivalent and CA1050 is the standard rule recognized across the .NET ecosystem, **MA0047 should be deprecated** in favor of CA1050.

### Migration Path

1. Enable **CA1050** in your analyzer configuration:
   ```xml
   <WarningsAsErrors Include="CA1050" />
   ```
2. Disable **MA0047**:
   ```xml
   <WarningsNotAsErrors Include="MA0047" />
   ```
3. Update any rule suppressions from MA0047 to CA1050

The rules are identical in name and functionality, making this deprecation a clear choice to reduce duplication and align with Microsoft's official analyzer ecosystem.

---
*Part of the broader initiative tracked in #835 to deprecate rules now covered by built-in .NET analyzers.*

**Comment by Meir017** [comment]
This issue should be closed - MA0004 similarity is already documented in [docs/comparison-with-other-analyzers.md](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/comparison-with-other-analyzers.md) in the "Similar rules" section - MA0004 is similar to CA2007 with the note that "MA does not report when the context is needed (WPF, WinForms, Blazor, etc.)"
